beacon/goclient/attest.go

```diff
@@ -324,53 +333,109 @@ func (gc *GoClient) fetchWeightedAttestationData(ctx context.Context,
+
 	// Initial score is based on height of source and target epochs.
 	score := float64(attestationData.Source.Epoch + attestationData.Target.Epoch)
+	logger.
+		With(zap.Float64("base_score", score)).
+		Info("base score was set. Fetching slot for block root")
 
-	// Increase score based on the nearness of the head slot.
-	slot, err := gc.blockRootToSlot(attestationData.BeaconBlockRoot, logger)
-	if err != nil {
-		logger.
-			With(zap.Error(err)).
-			Error("failed to obtain slot for block root")
-	} else {
-		score += float64(1) / float64(1+attestationData.Slot-slot)
-	}
+	ctx, cancel := context.WithTimeout(ctx, gc.weightedAttestationDataSoftTimeout/2)
+	defer cancel()
 
-	logger.With(
-		zap.Uint64("head_slot", uint64(slot)),
-		zap.Uint64("source_epoch", uint64(attestationData.Source.Epoch)),
-		zap.Uint64("target_epoch", uint64(attestationData.Target.Epoch)),
-		zap.Float64("score", score),
-	).Debug("scored attestation data")
+	ticker := time.NewTicker(time.Millisecond * 100)
+	defer ticker.Stop()
 
-	return score
-}
+	var (
+		retries uint8
+		start   = time.Now()
+	)
 
-func (gc *GoClient) blockRootToSlot(root phase0.Root, logger *zap.Logger) (phase0.Slot, error) {
-	slot, err, _ := gc.blockRootToSlotReqInflight.Do(root, func() (phase0.Slot, error) {
-		cacheResult := gc.blockRootToSlotCache.Get(root)
-		if cacheResult != nil {
-			cachedSlot := cacheResult.Value()
+	for {
+		slot, err := gc.blockRootToSlot(ctx, client, attestationData.BeaconBlockRoot, logger)
+		if err == nil {
+			// Increase score based on the nearness of the head slot.
+			score += float64(1) / float64(1+attestationData.Slot-slot)
+			logger.With(
+				zap.Duration("elapsed", time.Since(start)),
+				zap.Uint64("head_slot", uint64(slot)),
+				zap.Uint64("source_epoch", uint64(attestationData.Source.Epoch)),
+				zap.Uint64("target_epoch", uint64(attestationData.Target.Epoch)),
+				zap.Float64("score", score),
+			).Debug("scored attestation data")
+
+			return score
+		}
+
+		logger.
+			With(zap.Error(err)).
+			Warn("failed to obtain slot for block root")
+		select {
+		case <-ctx.Done():
 			logger.
-				With(zap.Uint64("cached_slot", uint64(cachedSlot))).
-				With(zap.Int("cache_len", gc.blockRootToSlotCache.Len())).
-				Debug("obtained slot from cache")
-			return cachedSlot, nil
+				With(zap.Uint8("retry_count", retries)).
+				With(zap.Duration("total_elapsed", time.Since(start))).
+				Error("timeout for obtaining slot for block root was reached. Returning base score")
+			return score
+		case <-ticker.C:
+			retries++
+			logger.
+				With(zap.Uint8("retry_count", retries)).
+				With(zap.Duration("total_elapsed", time.Since(start))).
+				Warn("retrying to obtain slot for block root")
 		}
+	}
+}
 
-		logger.Debug("slot was not found in cache, returning: '0'")
+func (gc *GoClient) blockRootToSlot(ctx context.Context, client Client, root phase0.Root, logger *zap.Logger) (phase0.Slot, error) {
+	cacheResult := gc.blockRootToSlotCache.Get(root)
```

Comment:
every log in the flow should have the block root
```suggestion
	logger := logger.With(fields.BlockRoot(root))
	cacheResult := gc.blockRootToSlotCache.Get(root)
```